### DISCUSSION TITLE
🤔 Treat pydantic model in function param as typed dict

### DISCUSSION THREAD

**Discussion by jd-solanki** [pydantic/pydantic#10527] [Question] [answered]
Hi 👋🏻 

I'm influenced by options pattern in [VueUse](https://vueuse.org/shared/useCounter/#usage-with-options). This allows seperating primary function param secondary function param while executing function. Assume we have `calculate` function that takes `a` & `b`  and by default it performs `addition` and we allow modifying this via dict along with other secondary params:
```py
calculate(1, 2)
calculate(1, 2, {"operation": "subtraction"})
calculate(1, 2, {"print_result": True})
calculate(1, 2, {})
```

While using pydantic best I came up with is below:
```py
from pydantic import BaseModel


class Options(BaseModel):
    operation: str | None = "addition"
    print_result: bool | None = False


def calculate(a: int, b: int, options: Options = Options()):
    print(f"options: {options}")
    print("---")


calculate(1, 2)
calculate(1, 2, Options(operation="subtraction"))
calculate(1, 2, Options(print_result=True))
calculate(1, 2, Options())
```

However, I need `Options` model everywhere I call calculate function. Wouldn't it be possible to just pass the dict instead of model instance like this with default values:
```py
calculate(1, 2)
calculate(1, 2, {"operation": "subtraction"})
calculate(1, 2, {"print_result": True})
calculate(1, 2, {})
```

**Comment by Viicos** [comment, marked as answer]
You can use `@validate_call` to do so:

```python
from pydantic import validate_call

...

@validate_call
def calculate(a: int, b: int, options: Options = Options()):
    print(f"options: {options}")
    print("---")

calculate(1, 2, {"operation": "subtraction"})
options: operation='subtraction' print_result=False
```

**Reply by jd-solanki** [reply]
We get type error with this:
![image](https://github.com/user-attachments/assets/afe17cbd-c90c-4dac-a325-fda0b03a475a)

**Reply by Viicos** [reply]
Yes, type checkers aren't aware of such a feature.
